### ISSUE TITLE
adaptive cluster doesn't scale down properly

### ISSUE THREAD

**Issue by danpf** [dask/dask-jobqueue#190]
I have a bunch of tasks, I ask for a SLURMCluster that's adaptive between 1-200 workers.  Occasionally one job will take really long.  I think adaptive should be aware that there's only n tasks left, and that it should only need n workers.  However instead dask keeps submitting to my scheduler (see the image below) and then the workers start, and immediately die because there's no jobs for them.

https://github.com/dask/distributed/issues/2329 Seems related, but it looks like jobqueue has its own adaptive implementation, so i'm not sure if that would fix anything

![image](https://user-images.githubusercontent.com/16297104/48143154-0c409980-e263-11e8-85a3-5d4e320c3f1c.png)

**Comment by guillaumeeb** [comment]
This seems definitely linked to https://github.com/dask/distributed/issues/2329.

Currently, dask-jobqueue still relies on distributed.deploy.adaptive, it's just separated from distributed.deploy.cluster. It may change in the future, but for the time being, if you can try the PR associated to the distributed issue, that should fix this irritating problem!

**Comment by danpf** [comment]
@guillaumeeb I've been using that since you suggested it, and I think that's not the problem.

The problem -- in my opinion --  is that jobqueue workers should be valued higher than simple dask-distributed workers.  In its current implementation submissions / cancellations are constantly made to the scheduler, but scheduler time valuable and slow to start, so it shouldn't just be stopped whenever it isn't using it.

On my scheduler (slurm) here's (see below) a few lines of random debugging code that I threw in around distributed and jobqueue.

In a few loops, which happen very quickly, the scheduler is scaled down from 15 -> 7, and then up to 15-> and then down to 11.

I think this is slightly due to the line in `get_scale_up_kwargs` (in distributed/deploy/adaptive)
```
target = math.ceil(self.scheduler.total_occupancy /self.target_duration)

````
the problem is (i think) twofold (please correct me if necessary):
1. target_duration (set by distributed/deploy/adaptive) is per-scheduler(Client) rather than on a per-submit(task) so if you have long / short jobs you're hurt
2. total_occupancy seems to vary a lot,  what's that? 

A possible solution, jobqeue workers should be to have workers 'almost unkillable' until they have accumulated a certain amount of time of having no tasks.

```
scale up kwargs: 15 len workers 2 min 0
scale up kwargs instance: 15
inst max: 15 40
scaling up? {'status': 'up'} {'n': 15}
scale up kwargs: 15 len workers 2 min 0
scale up kwargs instance: 15
inst max: 15 40
adapt sending up! {'n': 15}
should scale up = false
should scale up False
workers to close p set()
needs cpu/mem True False
should scale up True
workers to close p set()
scale up kwargs: 7 len workers 2 min 0
scale up kwargs instance: 7
inst max: 7 40
scaling up? {'status': 'up'} {'n': 7}
scale up kwargs: 7 len workers 2 min 0
scale up kwargs instance: 7
inst max: 7 40
adapt sending up! {'n': 7}
in scale up found ntc < count_pend
act and pend 15 n 7 ntc 8 pending 14
Stopping jobs: %s ['40903832', '40903833', '40903834', '40903835', '40903836', '40903837']
should scale up = false
should scale up False
workers to close p set()
needs cpu/mem True False
should scale up True
workers to close p set()
scale up kwargs: 15 len workers 8 min 0
scale up kwargs instance: 15
inst max: 15 40
scaling up? {'status': 'up'} {'n': 15}
scale up kwargs: 15 len workers 8 min 0
scale up kwargs instance: 15
inst max: 15 40
adapt sending up! {'n': 15}
should scale up = false
should scale up False
workers to close p set()
needs cpu/mem True False
should scale up True
workers to close p set()
scale up kwargs: 11 len workers 8 min 0
scale up kwargs instance: 11
inst max: 11 40
scaling up? {'status': 'up'} {'n': 11}
scale up kwargs: 11 len workers 8 min 0
scale up kwargs instance: 11
inst max: 11 40
adapt sending up! {'n': 11}
in scale up found ntc < count_pend
act and pend 15 n 11 ntc 4 pending 11
Stopping jobs: %s ['40903831', '40903839', '40903840', '40903841', '40903842', '40903843', '40903844']
```

**Comment by guillaumeeb** [comment]
> EDIT: I should mention, I thought the problem was that only N-tasks were available at the end, (which may still be the case) but after more investigation I think it might be something else! (or maybe something else completely different!)

It would be worth identifying what is the real problem here. I'm not sure I follow everything.

In your workflow, you don't only have long running tasks that remains at the end of the computation? The number of submitted and so pending tasks is varying a lot accross time?

>A possible solution, jobqeue workers should be to have workers 'almost unkillable' until they have accumulated a certain amount of time of having no tasks.

You could probably use `wait_count` kwarg for that. Worker actually get stopped if they're caught up `wait_count` times doing nothing (defaulting to 3), so after 3 seconds by default. Just increase `wait_count` to 300 for 5 minutes.

Some clarifications, I'm not sure it is clear currently:
-  `target_duration` is the duration you expect your whole computation to last. Adaptive will try to scale workers so that for each tasks submission, the client only wait by default 5 seconds before obtaining all results.
- `total_occupancy` is the Scheduler estimation of the total cpu time needed for computing all the pending tasks. The Scheduler tries to cleverly compute the an average task duration for updating this attribute. But in case there is a lot of variation of time between all individual tasks, I fear the estimation is hard to make and so total_occupancy can vary too much accross time.

**Comment by danpf** [comment]
Thanks for your response.

my workflow looks somewhat like:
where some of it is python, and some of it is `subprocess.call` to c++ binaries 

1. for x in n_inputs ->
2. Prepare inputs ~1000 small 5-20 second things
3. run initial task (1-10 minutes)*1
4. for x in n_from_[3] run  (~30-120 min)*(4-20)
5. compile [4] task results (5-20 min)*1
6.  for x in n_from_[5] (1hr to 3 hours)*(4-20)
7. compile 4th task results (5-10 min)*1
8. analysis & optimization -> ~100 small tasks (5-20 seconds)


I'll sort of just summarize what I've found out in the past few days debugging this:

- `wait_count:400` + `target_duration:'3600s'` appears to have fixed workers dying early in the protocol

- the PR you referenced appears to have fixed workers being submitted at the end when there are no tasks left

- I ran into ANOTHER problem, SLURM didn't seem to like being submitted a bunch of  batch scripts at the same time from a non-head node. Supposedly SLURM should be robust enough to handle that, but slurm would fail with 'resource temporarily unavailable' I fixed this by adding a short 0.3 second delay between submitting jobs to the scheduler. ( has anyone ever reported/noticed that before?)

I guess my only request at the end of the day would be for a way to set time per job in a `client.submit(` call instead of making adaptive assume that all of my jobs are going to take 1 hour.

Thank you for your help!

**Comment by guillaumeeb** [comment]
> target_duration:'3600s'
> I guess my only request at the end of the day would be for a way to set time per job in a client.submit( call instead of making adaptive assume that all of my jobs are going to take 1 hour.

I am under the impression that you are not intepreting `target_duration` correctly. It does not specify your job average duration, or tell dask how much time a task will last. It just configures how much time youwant to way before the computation ends. Dask will by itself estimate and refine your tasks duration.

Thus, I believe you don't need to set this value, you should try just setting `wait_count`.

> I ran into ANOTHER problem, SLURM didn't seem to like being submitted a bunch of batch scripts at the same time from a non-head node. Supposedly SLURM should be robust enough to handle that, but slurm would fail with 'resource temporarily unavailable' I fixed this by adding a short 0.3 second delay between submitting jobs to the scheduler. ( has anyone ever reported/noticed that before?)

This looks quite annoying, maybe you should talk to your cluster admins, or open an issue here to look for information, or on the Slurm issue tracker if there is one.
